go/vt/orchestrator/logic/tablet_discovery.go

```diff
@@ -285,21 +311,32 @@ func setReplicationSource(ctx context.Context, replica *topodatapb.Tablet, prima
 	return tmc.SetReplicationSource(ctx, replica, primary.Alias, 0, "", true, semiSync)
 }
 
-// shardPrimary finds the primary of the given keyspace-shard by reading the topo server
-func shardPrimary(ctx context.Context, keyspace string, shard string) (primary *topodatapb.Tablet, err error) {
-	si, err := ts.GetShard(ctx, keyspace, shard)
-	if err != nil {
-		return nil, err
-	}
-	if !si.HasPrimary() {
-		return nil, fmt.Errorf("no primary tablet for shard %v/%v", keyspace, shard)
-	}
-	// TODO(GuptaManan100): Instead of another topo call, use the local information by calling
-	// ReadTablet. Currently this isn't possible since we only have the primary alias and not the source host and port
-	// This should be fixed once the tablet alias is changed to be the primary key of the table
-	primaryInfo, err := ts.GetTablet(ctx, si.PrimaryAlias)
-	if err != nil {
-		return nil, err
+// shardPrimary finds the primary of the given keyspace-shard by reading the orchestrator backend
+func shardPrimary(keyspace string, shard string) (primary *topodatapb.Tablet, err error) {
+	query := `SELECT
```

Comment:
general comment: should we not put all query execution in some retryable template function?

Reply:
Yes, that would be a good addition, but so far not really needed, because even if the read fails, we just fail the recovery and then retry later.